### ISSUE TITLE
Reader failing with TruncatedDataException

### ISSUE THREAD

**Issue by jkhalack** [pravega/pravega#2664]
**Problem description**
On streams with dynamic scaling where the number of routing keys (20) is less than the average number of active segments (so some of the segments are empty) the readers are sometimes failing with `TruncatedDataException`.

For the `smallScaleSlow` stream shown in the segment heat chart below, there is a writer group of three writers and a reader group of one TAIL reader. 
At 3 hours 18 minutes after the start of our longevity test (that would be 12:17 chart time and 15:17 log time) the only reader finishes with `io.pravega.client.stream.TruncatedDataException`:
```
2018-06-05 15:17:35,137 11683422 [pool-8-thread-1] INFO  c.e.n.h.t.w.readers.PravegaReader - Removed Reader HulkSmallScaleSlow278f1654ead84e5999417823161e9505 from ReaderGroup HulkSmallScaleSlow at position hulk/smallScaleSlow/129=957719, hulk/smallScaleSlow/136=680764, hulk/smallScaleSlow/158=450590, hulk/smallScaleSlow/125=644602, hulk/smallScaleSlow/148=0, hulk/smallScaleSlow/115=1457960, hulk/smallScaleSlow/159=443702, hulk/smallScaleSlow/127=0, hulk/smallScaleSlow/128=668423, hulk/smallScaleSlow/132=707455, hulk/smallScaleSlow/113=0, hulk/smallScaleSlow/79=0, hulk/smallScaleSlow/173=99876, hulk/smallScaleSlow/174=112504, hulk/smallScaleSlow/152=702576
2018-06-05 15:17:35,143 11683428 [pool-8-thread-1] ERROR c.e.n.h.t.w.r.ReaderWorkerManager - Reader Finished with Error
io.pravega.client.stream.TruncatedDataException: null
	at io.pravega.client.stream.impl.EventStreamReaderImpl.handleSegmentTruncated(EventStreamReaderImpl.java:234)
	at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:110)
	at com.emc.nautilus.hulk.taskexecutor.workers.readers.PravegaReader.start(PravegaReader.java:96)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
```

Before this error there are a few attempts to read from segment `hulk/smallScaleSlow/79`:
```
2018-06-05 14:13:32,086 7840371 [pool-8-thread-1] INFO  i.p.c.s.impl.EventStreamReaderImpl - EventStreamReaderImpl( id=HulkSmallScaleSlow278f1654ead84e5999417823161e9505) acquiring segments {hulk/smallScaleSlow/79=0, hulk/smallScaleSlow/80=0}

...

2018-06-05 15:17:33,899 11682184 [epollEventLoopGroup-5-11] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Closing connection to segment hulk/smallScaleSlow/79 with exception: {}
io.pravega.shared.protocol.netty.ConnectionFailedException: null
	at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl$ResponseProcessor.connectionDropped(AsyncSegmentInputStreamImpl.java:60)
	at io.pravega.client.netty.impl.ClientConnectionInboundHandler.channelUnregistered(ClientConnectionInboundHandler.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelUnregistered(AbstractChannelHandlerContext.java:181)
...
2018-06-05 15:17:33,900 11682185 [epollEventLoopGroup-5-11] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Connection failed due to a io.pravega.shared.protocol.netty.ConnectionFailedException. Read requests will be retransmitted.
2018-06-05 15:17:33,900 11682185 [epollEventLoopGroup-5-11] WARN  i.p.c.s.i.AsyncSegmentInputStreamImpl - Exception while reading from Segment : hulk/smallScaleSlow/79
io.pravega.shared.protocol.netty.ConnectionFailedException: null
	at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl$ResponseProcessor.connectionDropped(AsyncSegmentInputStreamImpl.java:60)
	at io.pravega.client.netty.impl.ClientConnectionInboundHandler.channelUnregistered(ClientConnectionInboundHandler.java:86)
...
2018-06-05 15:17:33,949 11682234 [epollEventLoopGroup-5-11] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Received noSuchSegment No such segment: hulk/smallScaleSlow/79
2018-06-05 15:17:33,950 11682235 [epollEventLoopGroup-5-11] WARN  i.p.c.s.i.AsyncSegmentInputStreamImpl - Exception while reading from Segment : hulk/smallScaleSlow/79
io.pravega.client.segment.impl.SegmentTruncatedException: Segment no longer exists.
	at io.pravega.client.segment.impl.AsyncSegmentInputStreamImpl$ResponseProcessor.noSuchSegment(AsyncSegmentInputStreamImpl.java:73)
	at io.pravega.shared.protocol.netty.WireCommands$NoSuchSegment.process(WireCommands.java:241)
...
2018-06-05 15:17:34,789 11683074 [pool-8-thread-1] INFO  i.p.c.s.impl.EventStreamReaderImpl - EventStreamReaderImpl( id=HulkSmallScaleSlow278f1654ead84e5999417823161e9505) encountered truncation for segment hulk/smallScaleSlow/79 
...
2018-06-05 15:17:34,819 11683104 [epollEventLoopGroup-5-12] INFO  i.p.c.s.i.SegmentMetadataClientImpl - Closing connection as a result of receiving: No such segment: hulk/smallScaleSlow/79
2018-06-05 15:17:34,819 11683104 [epollEventLoopGroup-5-12] INFO  i.p.client.netty.impl.RawClient - Closing connection with exception: null
2018-06-05 15:17:34,827 11683112 [epollEventLoopGroup-5-12] INFO  i.p.client.netty.impl.RawClient - Closing connection with exception: null
2018-06-05 15:17:34,827 11683112 [pool-8-thread-1] INFO  i.p.c.s.impl.EventStreamReaderImpl - EventStreamReaderImpl( id=HulkSmallScaleSlow278f1654ead84e5999417823161e9505) encountered end of segment hulk/smallScaleSlow/79 
2018-06-05 15:17:34,827 11683112 [pool-8-thread-1] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Closing reader for hulk/smallScaleSlow/79
2018-06-05 15:17:34,827 11683112 [pool-8-thread-1] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Closing connection to segment: hulk/smallScaleSlow/79
2018-06-05 15:17:34,827 11683112 [pool-8-thread-1] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Connection failed due to a io.pravega.client.stream.impl.ConnectionClosedException. Read requests will be retransmitted.
2018-06-05 15:17:34,827 11683112 [epollEventLoopGroup-5-11] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Closing connection to segment: hulk/smallScaleSlow/79
2018-06-05 15:17:34,827 11683112 [epollEventLoopGroup-5-11] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Connection failed due to a io.pravega.shared.protocol.netty.ConnectionFailedException. Read requests will be retransmitted.
2018-06-05 15:17:34,886 11683171 [pool-8-thread-1] INFO  i.p.c.s.i.SegmentMetadataClientImpl - Closing segment metadata connection for hulk/smallScaleSlow/79
2018-06-05 15:17:34,886 11683171 [pool-8-thread-1] INFO  i.p.c.s.i.SegmentMetadataClientImpl - Closing connection with exception: null
2018-06-05 15:17:34,887 11683172 [pool-8-thread-1] ERROR c.e.n.h.t.w.readers.PravegaReader - Reader finished with Error
io.pravega.client.stream.TruncatedDataException: null
	at io.pravega.client.stream.impl.EventStreamReaderImpl.handleSegmentTruncated(EventStreamReaderImpl.java:234)
	at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:110)
	at com.emc.nautilus.hulk.taskexecutor.workers.readers.PravegaReader.start(PravegaReader.java:96)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
```
From the heat chart we can see that segment `79` was merged with `87` into new (empty) segment `106` at 11:10 AM chart time (14:10 log time). So it was assigned to the reader (at 14:13) only after it had been sealed. 

It must have been truncated at 14:47 (unfortunately I don't have controller logs with that entry), as the next truncation at 15:17 does not show segment `79` as being present (the descendant segment `106` is listed with offset 0):
```
2018-06-05 15:17:32,047 12351471 [controllerpool-26] INFO  i.p.c.s.e.r.TruncateStreamTask - Truncating stream hulk/smallScaleSlow at stream cut: {97=419020, 99=452599, 100=425047, 101=427630, 102=429352, 103=604422, 104=122836, 105=126280, 106=0, 107=121975, 108=129437, 109=244811, 110=106764, 111=117096, 112=335790, 113=0, 83=772030, 91=0, 94=0, 95=446859}
```

![giant-man smallScaleSlow segment heat chart](https://user-images.githubusercontent.com/25368596/40989434-215ea0f0-68c4-11e8-8397-4ab6c66df4e8.png)

[stdout.controller.giant-man.June5.log](https://github.com/pravega/pravega/files/2073177/stdout.controller.giant-man.June5.log)
[giant-man.smallScaleSlow.truncatedData.controller-server.June5.log](https://github.com/pravega/pravega/files/2073179/giant-man.smallScaleSlow.truncatedData.controller-server.June5.log)

Commit https://github.com/pravega/pravega/commit/6de4965871fd6b1767e0114b1a4768a9896f01cf was used to run the test.

**Problem location**
Client, reader

**Suggestions for an improvement**
The TAIL readers should not fail with `TruncatedDataException`.

**Comment by jkhalack** [comment]
Probably related to https://github.com/pravega/pravega/issues/2609

**Comment by jkhalack** [comment]
To possibly give more details, these are the logs for the same `smallScaleSlow` test (one TAIL reader, time limited data retention policy) with more recent pravega version and the client run in DEBUG mode.

Client log:
[mongoose.smallScaleSlow.controller_server_2018-06-11.22.log](https://github.com/pravega/pravega/files/2091212/mongoose.smallScaleSlow.controller_server_2018-06-11.22.log)

```
2018-06-11 17:29:39,780 9135576 [pool-8-thread-1] INFO  i.p.c.s.i.SegmentMetadataClientImpl - Closing segment metadata connection for hulk/smallScaleSlow/23.#epoch.12
2018-06-11 17:29:39,780 9135576 [pool-8-thread-1] INFO  i.p.c.s.i.SegmentMetadataClientImpl - Closing connection with exception: null
2018-06-11 17:29:39,781 9135577 [pool-8-thread-1] ERROR c.e.n.h.t.w.readers.PravegaReader - Reader finished with Error
io.pravega.client.stream.TruncatedDataException: null
	at io.pravega.client.stream.impl.EventStreamReaderImpl.handleSegmentTruncated(EventStreamReaderImpl.java:234)
	at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:110)
	at com.emc.nautilus.hulk.taskexecutor.workers.readers.PravegaReader.start(PravegaReader.java:96)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
```
From the stream heat chart, segment 23 ceased to exist 1 hour 15 minutes before the reader failure.

Pravega logs immediately before the failure:
[mongoose.smallScaleSlow.truncatedDataException.ed91a7f3-f6a0-4b64-9ecf-ef3ab0038802-1528738179956.zip]
(https://github.com/pravega/pravega/files/2091215/mongoose.smallScaleSlow.truncatedDataException.ed91a7f3-f6a0-4b64-9ecf-ef3ab0038802-1528738179956.zip)

**Comment by jkhalack** [comment]
Forgot to add that the above segment 23 was empty (no data written to it).

**Comment by andreipaduroiu** [comment]
@jkhalack Was there any sort of node failover during this test, especially immediately prior to this failing? 

#2678 seems to be reporting a rather similar issue, but I have pinpointed that to be due to a perfect storm of events that happen in a precise sequence, but they all involve a Segment Container failover.

Again, this may or may not be related, but once we merge the PR for 2678 in, we might want to see if we still encounter this.

**Comment by jkhalack** [comment]
Unfortunately, the cluster no longer exists, so there is no way to retrieve that information.
I'll look for segment store failures next time too.

**Comment by jadnhm** [comment]
This exception has just occurred on the vision cluster if anyone wants to have a look

**Comment by jadnhm** [comment]
[fd533884-c1a8-41ce-8ae8-697730b6c37b-1529414435010.zip](https://github.com/pravega/pravega/files/2115832/fd533884-c1a8-41ce-8ae8-697730b6c37b-1529414435010.zip)

**Comment by jkhalack** [comment]
Looks like this has been fixed in `v0.3.1` (I don't see it happening any more).
